### DISCUSSION TITLE
Where is the AR code?

### DISCUSSION THREAD

**Discussion by ConnollyLeon** [microsoft/Tutel#44] [Q&A]
I have read the [report](https://www.microsoft.com/en-us/research/blog/tutel-an-efficient-mixture-of-experts-implementation-for-large-dnn-model-training/) about tutel, and I am curious about the implementation of AlltoAll. 
I notice that tutel accelerate AlltoAll using CPU-GPU binding and adaptive routing tuning. I could find the CPU-GPU binding code in `system_init.py`, but fail to find code corresponding to  adaptive routing tuning. 

Could you please help me to locate  the code?

**Comment by abuccts** [comment]
Greetings! Adaptive Routing (AR) allows Azure NDv4 VM to automatically detect and avoid network congestion by dynamically selecting more optimal network paths in the InfiniBand networking. It requires cluster level configurations including InfiniBand switch etc. and cannot be controlled by application only. You can leverage its tuning if using Azure HPC image (which contains optimized environment and packages) with multiple NDv4 VMs. Please refer to [this article](https://techcommunity.microsoft.com/t5/azure-compute-blog/adaptive-routing-on-azure-hpc/ba-p/1205217) or Mellanox AR documents for more information about adaptive routing.
